translates: lifting-state-up (#30)
reactjs/nl.reactjs.org: content/docs/lifting-state-up.md

  - "docs/flux-todo-list.html"
---

> Vaak moeten verschillende componenten dezelfde veranderende gegevens weergeven.

Reviewer: ### Voorgestelde wijzigingen
- te veel line feeds
- state niet vertalen
- op te heffen => omhoog te tillen (is ook de titel)
- ancestor niet vertalen
- meest nabije => dichtstbijzijnde
- hun => zijn (enkelvoud state)

```suggestion
Vaak moeten verschillende componenten dezelfde veranderende gegevens weergeven. We raden aan de gedeelde state omhoog te tillen tot aan zijn dichtstbijzijnde gemeenschappelijke ancestor. Laten we eens kijken hoe dit in actie werkt.
```

### De tekst wordt dan
Vaak moeten verschillende componenten dezelfde veranderende gegevens weergeven. We raden aan de gedeelde state omhoog te tillen tot aan zijn dichtstbijzijnde gemeenschappelijke ancestor. Laten we eens kijken hoe dit in actie werkt.

### Oorspronkelijk
Vaak moeten verschillende componenten dezelfde veranderende gegevens weergeven.
We raden aan de gedeelde staat op te heffen tot aan hun meest nabije gemeenschappelijke voorvader.
Laten we eens kijken hoe dit in actie werkt.

> We raden aan de gedeelde staat op te heffen tot aan hun meest nabije gemeenschappelijke voorvader.

Reviewer: Deze vervalt dan
```suggestion
```

> Laten we eens kijken hoe dit in actie werkt.

Reviewer: Deze vervalt dan
```suggestion
```

In dit gedeelte zullen we een temperatuurcalculator maken die berekent of water bij een bepaalde temperatuur kookt.

> We zullen beginnen met een component genaamd `BiolingVerdict`.

Reviewer: ### Voorgestelde wijzigingen
- teveel enters
- `celcius` temperatuur => temperatuur `celcius`
- te koken => te laten koken
- het => die

```suggestion
We zullen beginnen met een component genaamd `BiolingVerdict`. Het accepteert de temperatuur `celsius` als een prop en drukt af of die voldoende is om water te laten koken:
```

### De tekst wordt dan
We zullen beginnen met een component genaamd `BiolingVerdict`. Het accepteert de temperatuur `celsius` als een prop en drukt af of die voldoende is om water te laten koken:

### Oorspronkelijk
We zullen beginnen met een component genaamd `BiolingVerdict`.
Het accepteert de `Celsius` temperatuur als een prop en drukt in de console af of het voldoende is om water te koken:

> Het accepteert de `Celsius` temperatuur als een prop en drukt in de console af of het voldoende is om water te koken:

Reviewer: Deze vervalt dan
```suggestion
```

}

> ```

Reviewer: In het voorbeeld zouden ook de afgedrukte teksten kunnen worden vertaald:

Het water zou koken.
Het water zou niet koken.

```
function BoilingVerdict(props) {
  if (props.celcius >= 100) {
    return <p>Het water zou koken.</p>
  }
  return <p>Het water zou niet koken.</p>;
}
```

Vervolgens zullen we een component genaamd `Calculator` maken.

> Het rendert een `<input>` waar je de temperatuur kunt invoeren en houdt deze waarde bij in `this.state.temperature`.

Reviewer: Twee zinnen naar 1 regel brengen

}

> ```

Reviewer: Tekst in de code

```
<legend>Voer de temperatuur in graden Celcius in:</legend>
```

[**Probeer het op CodePen**](https://codepen.io/gaearon/pen/ZXeOBm?editors=0010)

## Een Tweede Input Toevoegen {#adding-a-second-input}

Onze nieuwe vereiste is dat we, naast een Celsius input, ook een Fahrenheit input voorzien en deze worden gesynchroniseerd.

> We kunnen beginnen met een `TemperatureInput` component uit de `Calculator` af te zonderen.

Reviewer: af te zonderen => te halen
```suggestion
We kunnen beginnen met een `TemperatureInput` component uit de `Calculator` te halen.
```

> We zullen er een `scale` prop aan toevoegen die ofwel`"c"` ofwel `"f"` kan zijn.

Reviewer: Zin bij vorige voegen

}
```

[**Probeer het op CodePen**](https://codepen.io/gaearon/pen/jGBryx?editors=0010)

We hebben nu twee inputs, maar wanneer je een temperatuur ingeeft in één van beide, wordt de andere niet bijgewerkt.

> Dit is in tegenspraak met onze vereiste: we willen ze gesynchroniseerd houden.

Reviewer: onze => ons (het vereiste)
zin bij vorige voegen
```suggestion
Dit is in tegenspraak met ons vereiste: we willen ze gesynchroniseerd houden.
```

We kunnen ook de `BiolingVerdict` van `Calculator` niet weergeven.

> De `Calculator` kent de huidige temperatuur niet omdat deze verborgen is in `TemperatureInput`.

Reviewer: Zin bij vorige voegen

}
```

> Deze twee functies converteren nummers.

Reviewer: nummers => getallen
```suggestion
Deze twee functies converteren getallen.
```

> We zullen een andere functie schrijven die een string `temperature` en een conversiefunctie als argumenten heeft en een string teruggeeft.

Reviewer: een andere => nog een
```suggestion
We zullen nog een functie schrijven die een string `temperature` en een conversiefunctie als argumenten heeft en een string teruggeeft.
```

> We zullen deze gebruiken om de waarde van een input te berekenen op basis van een andere input.

Reviewer: basis van een => basis van de
Bovenstaande zinnen bij de vorige voegen
```suggestion
We zullen deze gebruiken om de waarde van een input te berekenen op basis van de andere input.
```

    // ...  
```

We willen echter deze twee inputs met elkaar gesynchroniseerd houden.

> Wanneer we de Celsius input bijwerken, moet de Fahrenheit input de geconverteerde temperatuur weerspiegelen en vice versa.

Reviewer: weerspiegelen => weergeven
Zin bij vorige voegen
```suggestion
Wanneer we de Celsius input bijwerken, moet de Fahrenheit input de geconverteerde temperatuur weergeven en vice versa.
```

In React wordt staat gedeeld door deze te verplaatsen naar de meest nabije gemeenschappelijke voorvader van de componenten die ze nodig hebben.
Dit wordt "staat omhoog tillen" genoemd.

> We zullen de lokale staat van de `TemperatureInput` verwijderen en ze verplaatsen naar de `Calculator`.

Reviewer: Twee zinnen bij die daarvoor voegen.

> Als de `Calculator` een gedeelde staat heeft, wordt het de "enige bron van waarheid" voor de huidige temperatuur voor beide inputs.

Reviewer: staat => state (niet vertalen)
```suggestion
Als de `Calculator` een gedeelde state heeft, wordt het de "enige bron van waarheid" voor de huidige temperatuur voor beide inputs.
```

Het kan er voor zorgen dat beiden waarden hebben die consistent zijn met elkaar.
Omdat de props van beide `TemperatureInput` componenten van dezelfde ouder `Calculator` afkomstig zijn, zullen de twee inputs altijd gesynchroniseerd zijn.

> Laten we eens kijken dit werkt stap voor stap.

Reviewer: Laten we eens stap voor stap kijken hoe dit werkt.
```suggestion
Laten we eens stap voor stap kijken hoe dit werkt.
```

Eerst zullen we `this.state.temperature` vervangen door `this.props.temperature` in de `TemperatureInput` component.

> Laten we voorlopig doen alsof `this.props.temperature` al bestaat, hoewel we het in de toekomst zullen moeten doorgeven aan de `Calculator`.

Reviewer: hoewel we het => hoewel we die
```suggestion
Laten we voorlopig doen alsof `this.props.temperature` al bestaat, hoewel we die in de toekomst zullen moeten doorgeven aan de `Calculator`.
```

    // ...
```

We weten dat [props read-only zijn](/doccs/components-and-props.html#props-are-read-only).

> Wanneer de `temperature` in de lokale staat zat, kon de `TemperatureInput` gewoon `this.setState()` aanroepen om ze te veranderen.

Reviewer: Wanneer => Toen
staat => state
```suggestion
Toen de `temperature` in de lokale state zat, kon de `TemperatureInput` gewoon `this.setState()` aanroepen om ze te veranderen.
```

Echter, nu dat de `temperature` afkomstig is uit de ouder als een prop, heeft de `TemperatureInput` er geen controle meer over.

> In React wordt dit doorgaans opgelost door de component "aangestuurd" te maken.

Reviewer: aangestuurd => controlled (niet vertalen)
```suggestion
In React wordt dit doorgaans opgelost door de component "controlled" te maken.
```

    // ...
```

>Opmerking:
>
>Er is geen speciale betekenis voor de namen van de `temperature` en `onTemperatureChange` props in aangepaste componenten. We hadden ze een andere naam kunnen geven, zoals `value` en `onChange`, wat een veel voorkomende conventie is.

> De `onTemperatureChange` prop zal samen met de `temperature` prop door de ouder `Calculator` worden voorzien.

Reviewer: voorzien => verzorgd
```suggestion
De `onTemperatureChange` prop zal samen met de `temperature` prop door de ouder `Calculator` worden verzorgd.
```

}
```

> Laten we nu kijken naar de `Calculator` component.

Reviewer: de => het

> We zullen de `temperature` en `scale` van de huidige input bijhouden in de lokale staat.

Reviewer: staat => state
```suggestion
We zullen de `temperature` en `scale` van de huidige input bijhouden in de lokale state.
```

> Dit is de staat die we "omhoog getild" hebben uit de inputs en deze zal dienen als de "bron van waarheid" voor beiden.

Reviewer: staat => state
```suggestion
Dit is de state die we "omhoog getild" hebben uit de inputs en deze zal dienen als de "bron van waarheid" voor beiden.
```

> Het is de minimale representatie van alle gegevens die we moeten kennen om beide ingangen weer te geven.

Reviewer: ingangen => inputs
```suggestion
Het is de minimale representatie van alle gegevens die we moeten kennen om beide inputs weer te geven.
```

> Als we bijvoorbeeld 37 invoeren in de Celsius input, is de staat van de `Calculator` component:

Reviewer: staat => state
```suggestion
Als we bijvoorbeeld 37 invoeren in de Celsius input, is de state van de `Calculator` component:
```

}
```

> Als we later het Fahrenheit veld veranderen in 212, is de staat van de `Calculator`:

Reviewer: staat => state
```suggestion
Als we later het Fahrenheit veld veranderen in 212, is de state van de `Calculator`:
```

}
```

> We hadden de waarde van beide inputs kunnen opslaan, maar dat blijkt overbodig. Het is voldoende om de waarde van de meest recent gewijzigde input en de schaal die deze vertegenwoordigt op te slaan. We kunnen dan de waarde van de andere input afleiden uit de huidige `temperature` en `scale`.

Reviewer: schaal => schaalverdeling
```suggestion
We hadden de waarde van beide inputs kunnen opslaan, maar dat blijkt overbodig. Het is voldoende om de waarde van de meest recent gewijzigde input en de schaalverdeling die deze vertegenwoordigt op te slaan. We kunnen dan de waarde van de andere input afleiden uit de huidige `temperature` en `scale`.
```

> De inputs blijven gesynchroniseerd omdat hun waarden worden berekend uit dezelfde staat:

Reviewer: staat => state
```suggestion
De inputs blijven gesynchroniseerd omdat hun waarden worden berekend uit dezelfde state:
```

}
```

[**Probeer het op CodePen**](https://codepen.io/gaearon/pen/WZpxpz?editors=0010)

Nu worden, ongeacht welke input je bewerkt, `this.state.temperature` en `this.state.scale` in de `Calculator` bijgewerkt.

> Eén van de inputs krijgt de waarde zoals deze is, dus elke gebruikersinvoer wordt behouden en de andere input waarde wordt altijd opnieuw berekend op basis van deze invoer.

Reviewer: dus elke => zodat elke
```suggestion
Eén van de inputs krijgt de waarde zoals deze is, zodat elke gebruikersinvoer wordt behouden en de andere input waarde wordt altijd opnieuw berekend op basis van deze invoer.
```

Laten we samenvatten wat er gebeurt als je een input bewerkt:

* React roept de functie aan die is gespecificeerd als `onChange` op de DOM `<input>`.
In ons geval is dit de `handleChange` methode in de component `TemperatureInput`.
* De `handleChange` methode in de `TemperatureInput` component roept `this.props.onTemperatureChange()` aan met de nieuwe gewenste waarde.

> De props, waaronder `onTemperatureChange`, werden voorzien door de bovenliggende component, de `Calculator`.

Reviewer: De props => Zijn props
```suggestion
Zijn props, waaronder `onTemperatureChange`, werden voorzien door de bovenliggende component, de `Calculator`.
```

> * Toen het eerder werd gerenderd, heeft de `Calculator` gespecificeerd dat `onTemperatureChange` van de Celsius `TemperatureInput` de `Calculator`'s `handleCelsiusChange` methode is en `onTemperatureChange` van de Fahrenheit `TemperatureInput` de `Calculator`'s `handleFahrenheitChange` methode.

Reviewer: Toen het => Toen hij
```suggestion
* Toen hij eerder werd gerenderd, heeft de `Calculator` gespecificeerd dat `onTemperatureChange` van de Celsius `TemperatureInput` de `Calculator`'s `handleCelsiusChange` methode is en `onTemperatureChange` van de Fahrenheit `TemperatureInput` de `Calculator`'s `handleFahrenheitChange` methode.
```

Dus één van beide methoden wordt aangeroepen afhankelijk van welke input bewerkt wordt.

> * Binnen deze methoden vraagt de `Calculator` component React om zichzelf opnieuw te renderen door `this.setState()` aan te roepen met de nieuwe input waarde en de huidige schaal van de input die we zojuist hebben bewerkt.

Reviewer: schaal van de => schaalverdeling die hoort bij de
```suggestion
* Binnen deze methoden vraagt de `Calculator` component React om zichzelf opnieuw te renderen door `this.setState()` aan te roepen met de nieuwe input waarde en de huidige schaalverdeling die hoort bij de input die we zojuist hebben bewerkt.
```

> * React roept de `render` methode van `Calculator` aan om te leren hoe de gebruiksersinterface eruit moet zien.

Reviewer: leren => weten
```suggestion
* React roept de `render` methode van `Calculator` aan om te weten hoe de gebruiksersinterface eruit moet zien.
```

De waarden van beide inputs worden opnieuw berekend op basis van de huidige temperatuur en de actieve schaal.
De temperatuurconversie wordt hier uitgevoerd.
* React roept de `render` methoden aan van de individuele `TemperatureInput` componenten met hun nieuwe props gespecificeerd door de `Calculator`.

> Het leert hoe hun gebruikersinterface eruit moet zien.

Reviewer: Hij leert => Hij komt erachter
```suggestion
Hij komt erachter hoe hun gebruikersinterface eruit moet zien.
```

* React roept de `render` methode van de `BoilingVerdict` component aan, waarbij de temperatuur in Celsius als prop wordt doorgegeven.
* React DOM werkt de DOM bij met het boiling verdict en zodat de gewenste input waarden getoond worden.
De input die we zojuist hebben bewerkt, ontvangt de huidige waarde en de andere input wordt bijgewerkt naar de temperatuur na de conversie.

Elke update gaat door deze zelfde stappen zodat de inputs gesynchroniseerd blijven.

> ## Les Geleerd {#lessons-learned}

Reviewer: Geleerde Lessen
```suggestion
## Geleerde Lessen {#lessons-learned}
```

> Er moet een enkele "bron van waarheid" zijn voor alle gegevens die veranderen in een React applicatie.

Reviewer: een enkele => één enkele
```suggestion
Er moet één enkele "bron van waarheid" zijn voor alle gegevens die veranderen in een React applicatie.
```

> Gewoonlijk wordt de staat eerst toegevoegd aan de component die deze nodig heeft voor zijn weergave.

Reviewer: staat => state
```suggestion
Gewoonlijk wordt de state eerst toegevoegd aan de component die deze nodig heeft voor zijn weergave.
```

> Als dan andere componenten ze ook nodig hebben, kan je ze omhoog tillen naar hun meest nabije gemeenschappelijke voorvader.

Reviewer: ze => hem (verwijst naar state) 2x
meest nabije => dichtstbijzijnde
voorvarder => ancestor
```suggestion
Als dan andere componenten hem ook nodig hebben, kan je hem omhoog tillen naar hun dichtstbijzijnde gemeenschappelijke ancestor.
```

> In plaats van de staat tussen verschillende componenten proberen te synchroniseren, zou je moeten vertrouwen op de [top-down data flow](/docs/state-and-lifecycle.html#the-data-flows-down).

Reviewer: staat => state
```suggestion
In plaats van de state tussen verschillende componenten proberen te synchroniseren, zou je moeten vertrouwen op de [top-down data flow](/docs/state-and-lifecycle.html#the-data-flows-down).
```

> Het tillen van staat houdt het schrijven van meer "boilerplate" code in dan bidirectionele bindingsbenaderingen, maar als voordeel, kost het minder moeite om bugs to vinden en te isoleren.

Reviewer: tillen => omhoog tillen
houdt ... in => brengt het schrijven van meer "boilerplate" code met zich mee
staat => state
bidirectionele bindingsbenaderingen = benaderingen met bi-directionele bindingen (wel knap gevonden!!)
bugs to => bugs te
```suggestion
Het omhoog tillen van state brengt met zich mee dat er meer "boilerplate" code geschreven moet worden dan bij benaderingen met bidirectionele bindings, maar het voordeel is dat  het minder moeite kost om bugs to vinden en te isoleren.
```

> Aangezien elke staat in een bepaalde component "leeft" en alleen die component ze kan veranderen, wordt de kans op fouten aanzienlijk verminderd.

Reviewer: staat => state
ze => hem
```suggestion
Aangezien elke state in een bepaalde component "leeft" en alleen die component hem kan veranderen, wordt de kans op fouten aanzienlijk verminderd.
```

> Bovendien kun je aangepaste logica implementeren om gebruikersinvoer te weigeren of te transformeren.

Reviewer: aangepaste logica => willekeurige aangepaste logica
```suggestion
Bovendien kun je willekeurige aangepaste logica implementeren om gebruikersinvoer te weigeren of te transformeren.
```

> Als iets kan afgeleid worden van props of state, zou het waarschijnlijk niet thuis horen in de state.

Reviewer: kan afgeleid => afgeleid kan
zou het  waarschijnlijk niet thuis horen in de state => hoort het waarschijnlijk niet thuis in de state
```suggestion
Als iets afgeleid kan worden van props of state, hoort het waarschijnlijk niet thuis in de state.
```

Bijvoorbeeld, in plaats van zowel `celsiusValue` als `fahrenheitValue` op te slaan, houden we enkel de laatst bewerkte `temperature` en de daarbij horende `scale` bij.

> De waarde van de andere input kan altijd worden berekend met de `render()` methode.

Reviewer: met de => in de
```suggestion
De waarde van de andere input kan altijd worden berekend in de `render()` methode.
```

> Dit laat ons toe om afronding op het andere veld aan en uit te zetten zonder enige precisie van de gebruikersinvoer te verliezen.

Reviewer: Dit laat ons toe om => Dit laat ons
```suggestion
Dit laat ons afronding op het andere veld aan en uit te zetten zonder enige precisie van de gebruikersinvoer te verliezen.
```

> Wanneer je iets verkeerd ziet in de gebruikersinterface, kun je [React Developer Tools](https://github.com/facebook/react-devtools) gebruiken om props te inspecteren en omhoog te gaan doorheen de boomstructuur totdat je de component vindt die verantwoordelijk is voor het bijwerken van de staat.

Reviewer: props => de props
doorheen de => door de
```suggestion
Wanneer je iets verkeerd ziet in de gebruikersinterface, kun je [React Developer Tools](https://github.com/facebook/react-devtools) gebruiken om de props te inspecteren en omhoog te gaan door de boomstructuur totdat je de component vindt die verantwoordelijk is voor het bijwerken van de staat.
```
